### DISCUSSION TITLE
Yarn 3 with private github package is always trying to lookup from registry.npmjs.org

### DISCUSSION THREAD

**Discussion by williamweckl** [renovatebot/renovate#11657] [Request Help] [answered]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Describe the bug

I'm trying to setup renovate to update my org private packages and it is not working. 

What I've tried so far:

- Create a .npmrc file at my repo, even using yarn. (tried with and without credentials, and with and without org scope)

```
@registry=https://npm.pkg.github.com
```

```
@myorg:registry=https://npm.pkg.github.com
```

- Set `"npmrc"` at renovate.json. (tried with and without credentials, and with and without org scope)

```json
{
  ...
  "encrypted": {
      "npmToken": "my-encrypted-token"
   },
  "npmrc": "registry=https://npm.pkg.github.com/\n//npm.pkg.github.com/:_authToken=${NPM_TOKEN}"
  ...
}
```

```json
{
  ...
  "encrypted": {
      "npmToken": "my-encrypted-token"
   },
  "npmrc": "myorg:registry=https://npm.pkg.github.com/\n//npm.pkg.github.com/:_authToken=${NPM_TOKEN}"
  ...
}
```

- "hostRules" at renovate.json.

```
"hostRules": [
  {
    "matchHost": "npm.pkg.github.com",
    "hostType": "npm",
    "encrypted": {
      "token": "encrypted-token"
    }
  }
],
```

- `npmScopes` at .yarnrc.yml

```
npmScopes:
  myorg:
    npmAlwaysAuth: true
    npmRegistryServer: 'https://npm.pkg.github.com/'
    npmAuthToken: 'TO_BE_REPLACED'
```

- `npmRegistries` at .yarnrc.yml

```
npmRegistries:
  '//npm.pkg.github.com/':
    npmAuthToken: 'TO_BE_REPLACED'
```

- `registryUrls` at renovate.json. (tried with and without "packageNames")

```
{
   ...
   "packageRules": [
      {
        "datasources": ["npm"],
        "packageNames": ["@myorg/mypkg"],
        "registryUrls": ["https://npm.pkg.github.com"]
      }
    ]
   ...
}
```

- A mix of all the strategies before mentioned and all together.

My package.json:

```json
{
   ...
   "name": "@myorg/package-that-im-setting-up-renovate",
   "private": true,
   "dependencies": {
      "@myorg/another-package-that-is-published-to-github-packages": "^0.0.65",
   }
   ...
}
```

Authentication is not the problem, I was not able to get so far, Renovate does not even try to reach out `npm.pkg.github.com`. It is always trying to lookup from `registry.npmjs.org`, that is weird since I'm using yarn. Renovate shouldn't try to update deps from `registry.yarnpkg.com`?

For obvious reason I've changed my org name and package name in the examples and logs.

Other thing to mention is that I'm commiting at  `renovate/configure` branch. The PR is still open and I didn't merge it. Is it possible that exists some cache that is preventing Renovate to get the changes I'm doing to that branch?

Thank you.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Dependency lookup failure: not found
{
  "err": {
    "name": "HTTPError",
    "timings": {
      "start": 1631153744016,
      "socket": 1631153744016,
      "lookup": 1631153745595,
      "connect": 1631153746245,
      "secureConnect": 1631153746537,
      "upload": 1631153746540,
      "response": 1631153746935,
      "end": 1631153746936,
      "phases": {
        "wait": 0,
        "dns": 1579,
        "tcp": 650,
        "tls": 292,
        "request": 3,
        "firstByte": 395,
        "download": 1,
        "total": 2920
      }
    },
    "message": "Response code 404 (Not Found)",
    "stack": "HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
    "options": {
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "accept": "application/json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://registry.npmjs.org/@myorg%2Fpkg-name",
      "hostType": "npm",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    },
    "response": {
      "statusCode": 404,
      "statusMessage": "Not Found",
      "body": {
        "error": "Not found"
      },
      "headers": {
        "date": "Thu, 09 Sep 2021 02:15:46 GMT",
        "content-type": "application/json",
        "content-length": "21",
        "connection": "close",
        "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
        "vary": "Accept-Encoding",
        "server": "cloudflare",
        "cf-ray": "68bce923ef60c97d-SEA"
      },
      "httpVersion": "1.1"
    }
  },
  "packageName": "@myorg/pkg-name"
}
DEBUG: Failed to look up dependency @myorg/pkg-name (@myorg/pkg-name)(packageFile="package.json", dependency="@myorg/pkg-name")
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
I appreciate you documenting all the things you tried, but there's too much going on in this to be an issue and so I've converted it to a Discussion until something narrower can be done, and a minimal reproduction repo made. The best solution for you would be for Renovate to read and understand Yarns own registry config in `.yarnrc.yml`, right? (https://yarnpkg.com/configuration/yarnrc#npmPublishRegistry)

**Reply by viceice** [reply]
We already use parts from `.yarnrc.yml`

https://github.com/renovatebot/renovate/blob/62a540c272d4f0b2738781d0442d61fd946cb482/lib/manager/npm/extract/yarn.ts#L45-L52
https://github.com/renovatebot/renovate/blob/b00cca0f1a7f3d21237525f9a518eed7ac58fd5f/lib/manager/npm/extract/index.ts#L125-L126

So should be easy to extract the registries and add them to extracted deps.
Maybe we need to allow custom registries for `npm` datasource

**Reply by williamweckl** [reply]
@rarkins Yes, I think that would be great. At CI I need just to add `npmScopes` to .yarnrc.yml with the proper npmRegistryServer and npmAuthToken. 
`npmScopes` is enough to local and CI environments to download the package versions. As I mentioned before, I've tried a lot of things, but `npmRegistries` is not really necessary.

Is there any reason that Renovate is not using my commited `.yarnrc.yml` file as is?

**Reply by rarkins** [reply]
It's just not implemented yet

**Reply by williamweckl** [reply]
i understand... The awkward thing is that yesterday I saw another issues closed that apparently was fixed with new versions of Renovate. Other people using yarn 2 and npmrc config, that somehow looks like worked for these other people.

https://github.com/renovatebot/renovate/issues/9985
https://github.com/renovatebot/renovate/issues/9537
https://github.com/renovatebot/renovate/issues/11100

With that I've came to the conclusion that probably I was doing something wrong... I really tried everything, including the workarounds mentioned in these issues. 

The information that is not implemented yet is new to me and I don't know if I'm happy because I'm not crazy or sad because I probably won't do what I need to do kkkk.

**Reply by rarkins** [reply]
#10052 added capabilities to write credentials from `hostRules` to `.yarnrc.yml`. Renovate doesn't currently *read* from `.yarnrc.yml` for registry information.

If you want Renovate to look up from an alternate host, the way to do that currently is a `.npmrc` in the repo or by setting the field `npmrc` in your config. Do you want every request looking up from GitHub's package registry or only your own scope?

**Reply by williamweckl** [reply]
I've tried the npmrc approach. Commited on the project and setting at config... It didn't work.
I want just the scoped package to look up from GitHub's package registry.

**Reply by williamweckl** [reply]
Since this is not implemented yet, would make sense to be an issue?

**Reply by rarkins** [reply]
If you have the time to create a reproduction repo, then please create a feature request for reading registry settings from `.yarnrc.yml`

**Comment by williamweckl** [comment, marked as answer]
For some reason, with the below settings it worked:

```json
{
  ...
  "force": {
    "constraints": {
      "yarn": "> 3"
    }
  },
  "hostRules": [
     {
      "matchHost": "https://npm.pkg.github.com/",
      "hostType": "npm",
      "encrypted": {
        "token": "*****************"
      }
    }
  ],
  "npmrc": "@myorg:registry=https://npm.pkg.github.com/"
}
```

Did it now undoing all the other things I've done before. I don't know if there was some cache or if the other things like having a .npmrc file at repo was somehow conflicting.

Forcing the yarn > 3 constraint also helped with other issues, at other repo node updates was causing artifact erros without it.

Since it is working we can close the discussion.

Thanks a lot for the support!
